File: src/powercontext/cli/transport.py
# Copyright (c) 2026 OceanBase.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Shared setup consent and non-secret per-host connection persistence."""

from __future__ import annotations

import json
import os
import re
import sys
from collections.abc import Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING, Any
from urllib.parse import urlsplit

import typer

from powercontext.client.transport_policy import (
    client_config_file,
    load_client_settings,
    parse_client_boolean,
    resolve_client_transport,
)
from powercontext.transport import is_loopback_host

if TYPE_CHECKING:
    from powercontext.cli.system import Diagnostic


@dataclass(frozen=True)
class SetupTransport:
    host: str
    server_url: str
    allow_insecure_http: bool


def is_remote_http(server_url: str) -> bool:
    """Classify plaintext endpoints without resolving DNS or making a request."""

    parsed = urlsplit(server_url)
    return parsed.scheme == "http" and not is_loopback_host(parsed.hostname)


def prepare_setup_transport(
    host: str,
    *,
    server_url: str | None = None,
    allow_insecure_http: bool | None = None,
    json_output: bool = False,
) -> SetupTransport:
    """Resolve and confirm the endpoint before any installation side effects."""

    from powercontext.cli.system import SetupError

    try:
        if host == "dsh":
            from powercontext.cli.native_transport import validate_dsh_setup_transport

            validate_dsh_setup_transport()
        prefix = "POWERCONTEXT_" + ("CLAUDE" if host == "claude-code" else host.upper().replace("-", "_")) + "_"
        has_environment_url = any(
            os.environ.get(key)
            for key in (
                prefix + "BASE_URL",
                prefix + "SERVER_URL",
                prefix + "ENDPOINT",
                "POWERCONTEXT_CLIENT_SERVER_URL",
            )
        )
        if server_url is None and not has_environment_url and not load_client_settings(host).get("server_url"):
            server_url = existing_native_endpoint(host)
        endpoint, allowed = resolve_client_transport(
            host, server_url=server_url, allow_insecure_http=allow_insecure_http
        )
        environment_consent = next(
            (
                os.environ[key]
                for key in (prefix + "ALLOW_INSECURE_HTTP", "POWERCONTEXT_CLIENT_ALLOW_INSECURE_HTTP")
                if key in os.environ
            ),
            None,
        )
        environment_refused = environment_consent is not None and not parse_client_boolean(environment_consent)
    except ValueError as error:
        raise SetupError(str(error)) from error
    endpoint = endpoint.removesuffix("/mcp").rstrip("/")
    if is_remote_http(endpoint):
        if not allowed:
            if (
                allow_insecure_http is None
                and not environment_refused
                and not json_output
                and sys.stdin.isatty()
                and typer.confirm(
                    f"{host}: {endpoint} uses unencrypted HTTP. Tokens and conversation content can be "
                    "read or modified in transit. Allow this endpoint?",
                    default=False,
                    err=True,
                )
            ):
                allowed = True
            else:
                raise SetupError(  # noqa: TRY003
                    "Remote HTTP requires explicit consent. Prefer HTTPS or an SSH tunnel, or use "
                    "--allow-insecure-http (POWERCONTEXT_CLIENT_ALLOW_INSECURE_HTTP=true)."
                )
        typer.echo(
            f"WARNING: {host} uses unencrypted HTTP at {endpoint}; credentials and content are not "
            "protected in transit. TLS verification and Server authentication remain unchanged.",
            err=True,
        )
    return SetupTransport(host, endpoint, allowed)


def existing_native_endpoint(host: str) -> str | None:
    """Preserve existing native endpoints when setup has no connection override."""

    if host != "workbuddy":
        from powercontext.cli.native_transport import _codex_url, _native_settings, _url

        if host == "codex":
            return _codex_url()
        native, url_key, _consent_key = _native_settings(host)
        return _url(native[url_key]) if url_key in native else None
    from powercontext.cli.workbuddy import workbuddy_home

    path = workbuddy_home() / "mcp.json"
    try:
        config = json.loads(path.read_text(encoding="utf-8"))
        value = config.get("mcpServers", {}).get("powercontext", {}).get("url")
        if isinstance(value, str):
            match = re.fullmatch(r"\$\{POWERCONTEXT_WORKBUDDY_SERVER_URL:-(.+)\}/mcp", value)
            return match.group(1) if match else value
    except (OSError, ValueError, AttributeError):
        pass
    return None


def save_setup_transport(
    settings: SetupTransport, *, native_updates: Sequence[tuple[Path, dict[str, Any]]] = ()
) -> None:
    """Save endpoint-bound consent, rolling back paired native writes on failure."""

    from powercontext.cli.system import SetupError, _write_bytes_atomically

    path = client_config_file()
    completed: list[tuple[Path, bytes | None]] = []
    try:
        config: Any = json.loads(path.read_text(encoding="utf-8")) if path.exists() else {"version": 1, "hosts": {}}
        if not isinstance(config, dict) or config.get("version") != 1 or not isinstance(config.get("hosts"), dict):
            raise ValueError("Client configuration must have version 1 and a hosts object")  # noqa: TRY003, TRY301
        config["hosts"][settings.host] = {
            "server_url": settings.server_url,
            "allow_insecure_http": settings.allow_insecure_http,
        }
        updates = list(native_updates)
        if settings.host == "hermes":
            native_path = hermes_config_file()
            native = json.loads(native_path.read_text(encoding="utf-8")) if native_path.exists() else {}
            if not isinstance(native, dict):
                raise ValueError("Hermes client configuration must be an object")  # noqa: TRY003, TRY301
            native.update(base_url=settings.server_url, allow_insecure_http=settings.allow_insecure_http)
            updates.append((native_path, native))
        updates.append((path, config))
        # Validate every destination before writing either half of a native/shared pair.
        writes = [
            (destination, destination.read_bytes() if destination.exists() else None, payload)
            for destination, payload in updates
        ]
        for destination, snapshot, payload in writes:
            _write_bytes_atomically(destination, (json.dumps(payload, indent=2) + "\n").encode())
            completed.append((destination, snapshot))
    except (OSError, ValueError) as error:
        for destination, snapshot in reversed(completed):
            try:
                if snapshot is None:
                    destination.unlink(missing_ok=True)
                else:
                    _write_bytes_atomically(destination, snapshot)
            except OSError:
                typer.echo(
                    f"WARNING: could not restore {destination}; check its configuration before restarting.", err=True
                )
        raise SetupError(f"Could not save PowerContext Client settings at {path}") from error  # noqa: TRY003


def hermes_config_file() -> Path:
    """Locate the same native configuration file used by the Hermes provider."""

    override = os.environ.get("POWERCONTEXT_HERMES_CONFIG")
    if override:
        return Path(override)
    return Path(os.environ.get("HERMES_HOME", str(Path.home() / ".hermes"))) / "powercontext" / "config.json"


def transport_diagnostic(host: str) -> Diagnostic:
    """Report explicitly insecure transport as degraded rather than silently green."""

    from powercontext.cli.native_transport import resolve_host_transport
    from powercontext.cli.system import Diagnostic, DiagnosticStatus

    try:
        endpoint, allowed = resolve_host_transport(host)
    except ValueError as error:
        return Diagnostic(status=DiagnosticStatus.FAILED, detail=str(error))
    if is_remote_http(endpoint):
        return Diagnostic(
            status=DiagnosticStatus.DEGRADED if allowed else DiagnosticStatus.FAILED,
            detail=(
                f"{endpoint}: insecure HTTP explicitly enabled; credentials and content are unencrypted"
                if allowed
                else f"{endpoint}: remote HTTP blocked; use HTTPS or explicitly allow insecure HTTP"
            ),
        )
    return Diagnostic(status=DiagnosticStatus.OK, detail=endpoint)


def add_transport_diagnostic(diagnostics: dict[str, Diagnostic], host: str) -> None:
    """Report the host's effective endpoint as well as transport policy failures."""

    diagnostic = transport_diagnostic(host)
    if not diagnostic.ok:
        diagnostics["transport"] = diagnostic
    else:
        from powercontext.cli.system import Diagnostic, DiagnosticStatus

        diagnostics["client_connection"] = Diagnostic(
            status=DiagnosticStatus.OK,
            detail=f"{diagnostic.detail}; client configuration: {client_config_file()}",
        )

File: integrations/openclaw/plugins/memory-powercontext/src/transport.ts
/*
 * Copyright (c) 2026 OceanBase.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { readFileSync } from 'node:fs'
import { homedir } from 'node:os'
import { join } from 'node:path'

// Each host is an independently distributed package; keep the shared client config
// contract and policy aligned with powercontext.client.transport_policy and powercontext.transport.
type SavedClient = { server_url?: string; allow_insecure_http?: boolean }

function optionalText(value: unknown): string | undefined {
  return typeof value === 'string' ? value.trim() || undefined : undefined
}

function optionalBoolean(value: unknown, name: string): boolean | undefined {
  if (value === undefined) return undefined
  if (typeof value !== 'boolean') throw new Error(`${name} must be a boolean`)
  return value
}

function environmentBoolean(env: NodeJS.ProcessEnv, name: string): boolean | undefined {
  if (env[name] === undefined) return undefined
  const value = env[name]!.trim().toLowerCase()
  if (['true', '1', 'yes', 'on'].includes(value)) return true
  if (['false', '0', 'no', 'off'].includes(value)) return false
  throw new Error(`${name} must be a boolean (true/false, 1/0, yes/no, on/off)`)
}

function readSavedClient(host: string, env: NodeJS.ProcessEnv): SavedClient {
  const home = optionalText(env.HOME) ?? homedir()
  const configuredPath = optionalText(env.POWERCONTEXT_CLIENT_CONFIG_FILE)
  const path = configuredPath?.startsWith('~/')
    ? join(home, configuredPath.slice(2))
    : configuredPath ?? join(home, '.config', 'powercontext', 'clients.json')
  let contents: string
  try {
    contents = readFileSync(path, 'utf8')
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === 'ENOENT') return {}
    throw new Error('Unable to read PowerContext client configuration', { cause: error })
  }
  let document: unknown
  try {
    document = JSON.parse(contents)
  } catch {
    throw new Error('PowerContext client configuration must be valid JSON')
  }
  if (!document || typeof document !== 'object' || Array.isArray(document)
    || (document as { version?: unknown }).version !== 1) {
    throw new Error('PowerContext client configuration must have version 1')
  }
  const hosts = (document as { hosts?: unknown }).hosts
  if (!hosts || typeof hosts !== 'object' || Array.isArray(hosts)) {
    throw new Error('PowerContext client configuration hosts must be an object')
  }
  const value = (hosts as Record<string, unknown>)[host]
  if (value === undefined) return {}
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error('PowerContext saved host configuration must be an object')
  }
  const entry = value as Record<string, unknown>
  if (entry.server_url !== undefined && !optionalText(entry.server_url)) {
    throw new Error('PowerContext saved server_url must be a non-empty string')
  }
  return {
    server_url: optionalText(entry.server_url),
    allow_insecure_http: optionalBoolean(entry.allow_insecure_http, 'allow_insecure_http'),
  }
}

export function normalizeServerUrl(value: string, allowInsecureHttp = false, name = 'PowerContext server URL'): string {
  optionalBoolean(allowInsecureHttp, 'allowInsecureHttp')
  let url: URL
  try {
    url = new URL(value)
  } catch {
    throw new Error(`${name} must be a valid HTTP(S) URL`)
  }
  if (!['http:', 'https:'].includes(url.protocol)) {
    throw new Error(`${name} must use HTTP or HTTPS`)
  }
  if (url.username || url.password || url.search || url.hash) {
    throw new Error(`${name} must not contain credentials, a query, or a fragment`)
  }
  const host = url.hostname.toLowerCase().replace(/^\[/, '').replace(/\]$/, '')
  const octets = host.split('.')
  const loopback = host === 'localhost' || host === '::1'
    || (octets.length === 4 && octets[0] === '127'
      && octets.every((octet) => /^\d{1,3}$/.test(octet) && Number(octet) <= 255))
  if (url.protocol === 'http:' && !loopback && !allowInsecureHttp) {
    throw new Error(`${name} must use HTTPS outside loopback; explicitly enable allow_insecure_http to permit plaintext HTTP`)
  }
  return url.toString().replace(/\/+$/, '').replace(/\/mcp$/, '').replace(/\/+$/, '')
}

export function resolveTransport(
  host: string,
  env: NodeJS.ProcessEnv,
  nativeUrl?: unknown,
  nativeConsent?: unknown,
  defaultUrl?: string,
): { baseUrl: string | undefined; allowInsecureHttp: boolean; source: 'environment' | 'plugin' | 'saved' | 'default' } {
  const prefix = `POWERCONTEXT_${host.toUpperCase()}`
  const saved = readSavedClient(host, env)
  const environmentUrl = optionalText(env[`${prefix}_BASE_URL`])
    ?? optionalText(env[`${prefix}_SERVER_URL`])
    ?? optionalText(env[`${prefix}_ENDPOINT`])
    ?? optionalText(env.POWERCONTEXT_CLIENT_SERVER_URL)
  const pluginUrl = optionalText(nativeUrl)
  const selectedUrl = environmentUrl ?? pluginUrl ?? saved.server_url ?? defaultUrl
  const normalized = selectedUrl === undefined ? undefined : normalizeServerUrl(selectedUrl, true, `${prefix}_BASE_URL`)
  const savedUrl = saved.server_url === undefined ? undefined : normalizeServerUrl(saved.server_url, true)
  const hostConsent = environmentBoolean(env, `${prefix}_ALLOW_INSECURE_HTTP`)
  const commonConsent = environmentBoolean(env, 'POWERCONTEXT_CLIENT_ALLOW_INSECURE_HTTP')
  const pluginConsent = optionalBoolean(nativeConsent, 'allowInsecureHttp')
  const nativeEndpoint = pluginUrl === undefined ? undefined : normalizeServerUrl(pluginUrl, true)
  // Persisted native consent belongs to its endpoint, just like setup-saved consent.
  // An explicit refusal remains effective even when the endpoint is overridden.
  const matchingNativeConsent = pluginConsent === false
    ? false
    : normalized !== undefined && normalized === nativeEndpoint ? pluginConsent : undefined
  const allowInsecureHttp = hostConsent ?? commonConsent ?? matchingNativeConsent
    ?? (normalized !== undefined && normalized === savedUrl ? saved.allow_insecure_http : undefined) ?? false
  return {
    baseUrl: normalized === undefined ? undefined : normalizeServerUrl(normalized, allowInsecureHttp, `${prefix}_BASE_URL`),
    allowInsecureHttp,
    source: environmentUrl ? 'environment' : pluginUrl ? 'plugin' : saved.server_url ? 'saved' : 'default',
  }
}
